lib/watcher.ts

```diff
@@ -15,20 +15,34 @@ interface WatcherOptions {
 // WatcherAdapter handles I/O via the sane package, and could be pluggable in
 // principle.
 
+type Deferred = {
+  promise: Promise<void>;
+  resolve: (value?: any) => void;
+  reject: (error?: any) => void;
+}
+
+function deferred() : Deferred {
+  const deferred = {} as Deferred;
+  deferred.promise = new Promise((resolve, reject) => {
+    deferred.resolve = resolve;
+    deferred.reject = reject ;
+  });
+
+  return deferred;
+}
+
 class Watcher extends EventEmitter {
-  _changedFiles: string[];
-  _quitting?: boolean; // is this ever set
-  _rebuildScheduled: boolean;
-  _ready: boolean;
-  _quittingPromise: Promise<void> | null;
-  _lifetime: {
-    promise?: Promise<void>;
-    resolve?: (value: any) => void;
-    reject?: (error: any) => void;
-  } | null;
+  private _changedFiles: string[] = [];
+  private _quitting?: boolean; // is this ever set
+  private _rebuildScheduled = false;
+  private _ready = false;
+  private _quittingPromise: Promise<void> | null = null;
+  private _lifetime: Deferred | null = null;
+  private _nextBuild? = deferred();
```

Comment:
TODO: there are lots of seemingly overlapping variables here, let's try to reduce this so it's easier to reason about